ptvsd/messaging.py

```diff
@@ -0,0 +1,292 @@
+    def __init__(self, stream, handlers=None):
+        self.send_callback = lambda channel, message: None
+        self.receive_callback = lambda channel, message: None
+        self._lock = threading.Lock()
+        self._stop = threading.Event()
+        self._stream = stream
+        self._seq_iter = itertools.count(1)
+        self._requests = {}
+        self._handlers = handlers
+        self._worker = threading.Thread(target=self._process_incoming_messages)
+        self._worker.daemon = True
+
+    def start(self):
+        self._worker.start()
+
+    def wait(self):
+        self._worker.join()
+
+    def _send_message(self, type, rest={}):
+        with self._lock:
+            seq = next(self._seq_iter)
+        message = {
+            'seq': seq,
+            'type': type,
+        }
+        message.update(rest)
+        with self._lock:
+            self._stream.write_json(message)
+        self.send_callback(self, message)
+        return seq
+
+    def send_request(self, command, arguments=None):
+        d = {'command': command}
+        if arguments is not None:
+            d['arguments'] = arguments
+        seq = self._send_message('request', d)
+        request = Request(self, seq)
+        with self._lock:
+            self._requests[seq] = request
+        return request
+
+    def send_event(self, event, body=None):
+        d = {'event': event}
+        if body is not None:
+            d['body'] = body
+        self._send_message('event', d)
+
+    def send_response(self, request_seq, success, command, error_message=None, body=None):
+        d = {
+            'request_seq': request_seq,
+            'success': success,
+            'command': command,
+        }
+        if success:
+            if body is not None:
+                d['body'] = body
+        else:
+            if error_message is not None:
+                d['message'] = error_message
+        self._send_message('response', d)
+
+    def on_message(self, message):
+        self.receive_callback(self, message)
+        seq = message['seq']
+        typ = message['type']
+        if typ == 'request':
+            command = message['command']
+            arguments = message.get('arguments', None)
+            self.on_request(seq, command, arguments)
+        elif typ == 'event':
+            event = message['event']
+            body = message.get('body', None)
+            self.on_event(seq, event, body)
+        elif typ == 'response':
+            request_seq = message['request_seq']
+            success = message['success']
+            command = message['command']
+            error_message = message.get('message', None)
+            body = message.get('body', None)
+            self.on_response(seq, request_seq, success, command, error_message, body)
+        else:
+            raise IOError('Incoming message has invalid "type":\n%r' % message)
+
+    def on_request(self, seq, command, arguments):
+        handler_name = '%s_request' % command
```

Comment:
Low pri: Should handler decide the name?

Reply:
You mean the suffix? I figured so long as there's a clear convention that guarantees no name clashes, it's fine. Just can't think of any practical scenario where the handler class would want to use custom names... but one possibility is to have a tiny handler wrapper that just has `request` and `event`, and then does its own dispatching to the real handler object by whatever rules it wants.